compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala

```diff
@@ -815,6 +815,7 @@ class OrderingConstraint(private val boundsMap: ParamBounds,
     newConstraint(hardVars = this.hardVars + tv)
 
   def instType(tvar: TypeVar): Type = entry(tvar.origin) match
+    case TypeBounds(lo, hi) if lo == hi => lo
```

Comment:
This is a very interesting observation. 
I encountered a similar problem when working on ##19955.

I believe the part of the code attempting to do this logic is 
https://github.com/scala/scala3/blob/63810dca83be96b29097acf6c554808932466d1a/compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala#L304-L311

Reply:
I found that this happens after we search for implicits and, in result, commit the new typerstate, in the version without the intersection type the constraint was consisting of a single type, so it was easily removed. For intersection types, we had that weird TypeBound with equal lower and upper bounds, so the constraint was kept and we later compiled with an unfulfilled type variable, causing the bad performance

Reply:
It's `TypeBounds(Repro.ObjectId & Endpoints.this.Foo, Repro.ObjectId & Endpoints.this.Foo)`, or `TypeBounds(AndType(TypeRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,module class <empty>)),module class Repro$)),trait ObjectId),TypeRef(ThisType(TypeRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,module class <empty>)),module class Repro$)),class Endpoints)),trait Foo)),AndType(TypeRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,module class <empty>)),module class Repro$)),trait ObjectId),TypeRef(ThisType(TypeRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,module class <empty>)),module class Repro$)),class Endpoints)),trait Foo)))`

The `eq` comparison in place of `==` is a sufficient fix for the reproduction in #20120, but I am wondering where this constraint (`TX1 >: Repro.ObjectId & Endpoints.this.Foo <: Repro.ObjectId & Endpoints.this.Foo`) is created (and will try to investigate that)